mgmt: mcumgr: transport: shell: smp compilation error

#63967 removed the struct declaration required by `smp_shell_input_timeout_handler`, this is currently causing the CI in Zephyr trunk to fail, it also causes compilation in our internal firmware build

**To Reproduce**
Steps to reproduce the behavior:
1. checkout `main`
2. ` twister -i -p nrf52840dk_nrf52840 -T zephyr/tests/subsys/mgmt/mcumgr/all_options/`

**Expected behavior**
Compilation should pass

**Impact**
showstopper

**Logs and console output**
```
zephyr/subsys/mgmt/mcumgr/transport/CMakeFiles/subsys__mgmt__mcumgr__transport.dir/src/smp_shell.c.obj -c /__w/zephyr/zephyr/subsys/mgmt/mcumgr/transport/src/smp_shell.c
/__w/zephyr/zephyr/subsys/mgmt/mcumgr/transport/src/smp_shell.c: In function 'smp_shell_input_timeout_handler':
/__w/zephyr/zephyr/subsys/mgmt/mcumgr/transport/src/smp_shell.c:68:53: error: invalid use of undefined type 'struct shell_uart'
   68 |         struct smp_shell_data *const data = &sh_uart->ctrl_blk->smp;
      |                                                     ^~
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain Zephyr SDK
 - Commit SHA or Version used: 56f73bde0fb086fae38cfb5afb2f2500a340b39b

--- a/include/zephyr/shell/shell_uart.h
+++ b/include/zephyr/shell/shell_uart.h
@@ -7,6 +7,7 @@
 #ifndef SHELL_UART_H__
 #define SHELL_UART_H__
 
+#include <zephyr/mgmt/mcumgr/transport/smp_shell.h>
 #include <zephyr/shell/shell.h>
 
 #ifdef __cplusplus
@@ -22,6 +23,13 @@ extern "C" {
  * @returns Pointer to the shell instance.
  */
 const struct shell *shell_backend_uart_get_ptr(void);
+
+/**
+ * @brief This function provides pointer to the smp shell data of the UART shell transport.
+ *
+ * @returns Pointer to the smp shell data.
+ */
+struct smp_shell_data *shell_uart_smp_shell_data_get_ptr(void);
 
 #ifdef __cplusplus
 }

--- a/subsys/mgmt/mcumgr/transport/src/smp_shell.c
+++ b/subsys/mgmt/mcumgr/transport/src/smp_shell.c
@@ -59,13 +59,10 @@ enum smp_shell_mcumgr_state {
 };
 
 #ifdef CONFIG_MCUMGR_TRANSPORT_SHELL_INPUT_TIMEOUT
-extern struct shell_transport shell_transport_uart;
-
 static void smp_shell_input_timeout_handler(struct k_timer *timer)
 {
 	ARG_UNUSED(timer);
-	struct shell_uart *sh_uart = (struct shell_uart *)shell_transport_uart.ctx;
-	struct smp_shell_data *const data = &sh_uart->ctrl_blk->smp;
+	struct smp_shell_data *const data = shell_uart_smp_shell_data_get_ptr();
 
 	atomic_clear_bit(&data->esc_state, ESC_MCUMGR_PKT_1);
 	atomic_clear_bit(&data->esc_state, ESC_MCUMGR_PKT_2);

--- a/subsys/shell/backends/shell_uart.c
+++ b/subsys/shell/backends/shell_uart.c
@@ -554,6 +554,10 @@ static int read(const struct shell_transport *transport,
 #ifdef CONFIG_MCUMGR_TRANSPORT_SHELL
 static void update(const struct shell_transport *transport)
 {
+	/*
+	 * This is dependent on the fact that `struct shell_uart_common`
+	 * is always the first member, regardless of the UART configuration
+	 */
 	struct shell_uart_common *sh_uart = (struct shell_uart_common *)transport->ctx;
 
 	smp_shell_process(&sh_uart->smp);
@@ -582,6 +586,15 @@ SHELL_DEFINE(shell_uart, CONFIG_SHELL_PROMPT_UART, &shell_transport_uart,
 	     CONFIG_SHELL_BACKEND_SERIAL_LOG_MESSAGE_QUEUE_SIZE,
 	     CONFIG_SHELL_BACKEND_SERIAL_LOG_MESSAGE_QUEUE_TIMEOUT,
 	     SHELL_FLAG_OLF_CRLF);
+
+#ifdef CONFIG_MCUMGR_TRANSPORT_SHELL
+struct smp_shell_data *shell_uart_smp_shell_data_get_ptr(void)
+{
+	struct shell_uart_common *common = (struct shell_uart_common *)shell_transport_uart.ctx;
+
+	return &common->smp;
+}
+#endif
 
 static int enable_shell_uart(void)
 {

